### PR TITLE
[MI-2199]: Made changes on server to add constraints on project/channel visibility on subscription cards.

### PR DESCRIPTION
**Pull request by avas27JTG**
Made changes on server to add constraints on project/channel visibility on subscription cards.

### CODE REVIEW

File: server/plugin/utils.go [hunk 1]
```diff
@@ -275,3 +275,35 @@ func (p *Plugin) GetOffsetAndLimitFromQueryParams(r *http.Request) (offset, limi
 
 	return page * limit, limit
 }
+
+func (p *Plugin) GetSubscriptionsForOnlyAccessibleChannelsOrProjects(subscriptionList []*serializers.SubscriptionDetails, teamID, mattermostUserID string) ([]*serializers.SubscriptionDetails, error) {
```
Comment (Kshitij-Katiyar) [inline server/plugin/utils.go:279]:
```suggestion
func (p *Plugin) GetSubscriptionsForAccessibleChannelsOrProjects(subscriptionList []*serializers.SubscriptionDetails, teamID, mattermostUserID string) ([]*serializers.SubscriptionDetails, error) {
```

File: server/plugin/utils.go [hunk 2]
```diff
@@ -275,3 +275,35 @@ func (p *Plugin) GetOffsetAndLimitFromQueryParams(r *http.Request) (offset, limi
...
+
+	projectList, err := p.Store.GetAllProjects(mattermostUserID)
+	if err != nil {
+		p.API.LogError(constants.ErrorFetchProjectList, "Error", err.Error())
+		return nil, err
+	}
+
+	var filteredSubscriptionList []*serializers.SubscriptionDetails
+	for _, subscription := range subscriptionList {
+		// For each subscripton on a project check if a user is an admin or member of the MM channel to return subscriptions
```
Comment (Kshitij-Katiyar) [inline server/plugin/utils.go:294]:
```suggestion
		// For each subscription on a project check if a user is an admin or member of the MM channel to return subscriptions
```

File: webapp/src/services/index.ts [hunk 3]
```diff
@@ -67,7 +67,7 @@ const azureDevOpsPluginApi = createApi({
         [Constants.pluginApiServiceConfigs.getSubscriptionList.apiServiceName]: builder.query<SubscriptionDetails[], FetchSubscriptionList>({
             query: (params) => ({
                 headers: {[Constants.common.HeaderCSRFToken]: Cookies.get(Constants.common.MMCSRF)},
-                url: Constants.pluginApiServiceConfigs.getSubscriptionList.path,
+                url: `${Constants.pluginApiServiceConfigs.getSubscriptionList.path}/${params.team_id}`,
```
Comment (Kshitij-Katiyar) [inline webapp/src/services/index.ts:70]: here we are sending the team id in path param as well as in body do we need to send this twice

File: server/constants/messages.go [hunk 4]
```diff
@@ -34,32 +34,36 @@ const (
...
+	ErrorFetchProjectList              = "Error in fetching project list"
+	ErrorDecodingBody                  = "Error in decoding body"
+	ErrorCreateTask                    = "Error in creating task"
+	ErrorLinkProject                   = "Error in linking the project"
+	FetchSubscriptionListError         = "Error in fetching subscription list"
+	FetchFilteredSubscriptionListError = "Error in fetching filtered subscription list"
+	CreateSubscriptionError            = "Error in creating subscription"
+	ErrorCheckingProjectAdmin          = "Error in checking if user is an admin on the project %s"
+	ProjectNotLinked                   = "Requested project is not linked"
+	GetSubscriptionListError           = "Error getting Subscription List"
```
Comment (raghavaggarwal2308) [inline server/constants/messages.go:54]:
```suggestion
	GetSubscriptionListError           = "Error getting subscription list"
```

File: server/plugin/api.go [hunk 5]
```diff
@@ -117,11 +117,24 @@ func (p *Plugin) handleLink(w http.ResponseWriter, r *http.Request) {
...
 	}
 
+	isAdmin := false
+	subscriptionStatusCode, subscriptionErr := p.Client.CheckIfUserIsProjectAdmin(body.Organization, response.ID, p.GetPluginURL(), mattermostUserID)
+	if subscriptionErr != nil {
+		if subscriptionStatusCode == http.StatusBadRequest && strings.Contains(subscriptionErr.Error(), fmt.Sprintf(constants.ErrorMessageForAdmin, constants.SubscriptionEventTypeDummy)) {
+			isAdmin = true
+		} else {
+			p.API.LogError(fmt.Sprintf(constants.ErrorCheckingProjectAdmin, body.Project), "Error", subscriptionErr.Error())
+			p.handleError(w, r, &serializers.Error{Code: statusCode, Message: constants.ErrorLinkProject})
```
Comment (raghavaggarwal2308) [inline server/plugin/api.go:127]:
```suggestion
			p.handleError(w, r, &serializers.Error{Code: subscriptionStatusCode, Message: constants.ErrorLinkProject})
```

File: server/plugin/api_test.go [hunk 6]
```diff
@@ -680,6 +682,12 @@ func TestHandleGetSubscriptions(t *testing.T) {
 			marshalError: errors.New("mockError"),
 			statusCode:   http.StatusInternalServerError,
 		},
+		{
+			description: "HandleGetSubscriptions: GetSubscriptionsForOnlyAccessibleChannelsOrProjects gives error",
```
Comment (raghavaggarwal2308) [inline server/plugin/api_test.go:686]:
```suggestion
			description: "HandleGetSubscriptions: GetSubscriptionsForAccessibleChannelsOrProjects gives error",
```

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

**Review by Kshitij-Katiyar — APPROVED** [review]

File: server/plugin/utils.go [hunk 7]
```diff
@@ -275,3 +275,35 @@ func (p *Plugin) GetOffsetAndLimitFromQueryParams(r *http.Request) (offset, limi
...
+	var filteredSubscriptionList []*serializers.SubscriptionDetails
+	for _, subscription := range subscriptionList {
+		// For each subscription on a project check if a user is an admin or member of the MM channel to return subscriptions
+		if projectDetails, isProjectLinked := p.IsProjectLinked(projectList, serializers.ProjectDetails{OrganizationName: subscription.OrganizationName, ProjectName: subscription.ProjectName}); isProjectLinked {
+			if projectDetails.IsAdmin {
+				filteredSubscriptionList = append(filteredSubscriptionList, subscription)
+			} else {
+				for _, channel := range channels {
+					if subscription.ChannelID == channel.Id {
+						filteredSubscriptionList = append(filteredSubscriptionList, subscription)
```
Comment (manojmalik20) [inline server/plugin/utils.go:301]:
```suggestion
						filteredSubscriptionList = append(filteredSubscriptionList, subscription)
						break
```

**Review by manojmalik20 — APPROVED** [review]

**Review by raghavaggarwal2308 — APPROVED** [review]

**Pull request merged**